File: src/algorithms/trees/advanced/huffman-coding-tree/sources/huffman-coding-tree.ts
// Huffman Coding Tree — build optimal prefix-free encoding from character frequencies

interface HuffmanNode {
  freq: number;
  char: string | null;
  left: HuffmanNode | null;
  right: HuffmanNode | null;
}

function huffmanCodingTree(frequencies: { char: string; freq: number }[]): Record<string, string> {
  // Build a min-heap (priority queue) from frequencies
  const minHeap: HuffmanNode[] = frequencies.map(({ char, freq }) => ({
    freq,
    char,
    left: null,
    right: null,
  })); // @step:initialize

  // Sort ascending to simulate a min-heap
  minHeap.sort((nodeA, nodeB) => nodeA.freq - nodeB.freq); // @step:select-min-freq

  while (minHeap.length > 1) {
    // Extract two minimums
    const leftNode = minHeap.shift()!; // @step:select-min-freq
    const rightNode = minHeap.shift()!; // @step:select-min-freq

    // Merge into a new internal node
    const merged: HuffmanNode = {
      freq: leftNode.freq + rightNode.freq,
      char: null,
      left: leftNode,
      right: rightNode,
    }; // @step:build-node

    // Re-insert maintaining sorted order
    const insertPos = minHeap.findIndex((node) => node.freq > merged.freq);
    if (insertPos === -1) minHeap.push(merged);
    else minHeap.splice(insertPos, 0, merged); // @step:build-node
  }

  const huffmanRoot = minHeap[0] ?? null;
  const encodings: Record<string, string> = {};

  function generateCodes(node: HuffmanNode | null, code: string): void {
    if (!node) return;
    if (node.char !== null) {
      encodings[node.char] = code || "0"; // @step:encode-char
      return;
    }
    generateCodes(node.left, code + "0"); // @step:traverse-left
    generateCodes(node.right, code + "1"); // @step:traverse-right
  }

  generateCodes(huffmanRoot, "");
  return encodings; // @step:complete
}

File: src/algorithms/trees/advanced/huffman-coding-tree/sources/HuffmanCodingTree.cpp
// Huffman Coding Tree — build optimal prefix-free encoding from character frequencies
#include <string>
#include <vector>
#include <map>
#include <algorithm>
using namespace std;

struct HuffmanNode {
    int freq;
    char charVal;
    bool isLeaf;
    HuffmanNode* left;
    HuffmanNode* right;
    HuffmanNode(char c, int f) : freq(f), charVal(c), isLeaf(true), left(nullptr), right(nullptr) {}
    HuffmanNode(int f, HuffmanNode* l, HuffmanNode* r) : freq(f), charVal(0), isLeaf(false), left(l), right(r) {}
};

void generateCodes(HuffmanNode* node, string code, map<char, string>& encodings) {
    if (!node) return;
    if (node->isLeaf) {
        encodings[node->charVal] = code.empty() ? "0" : code; // @step:encode-char
        return;
    }
    generateCodes(node->left, code + "0", encodings);  // @step:traverse-left
    generateCodes(node->right, code + "1", encodings); // @step:traverse-right
}

map<char, string> huffmanCodingTree(vector<pair<char, int>> frequencies) {
    vector<HuffmanNode*> minHeap;
    for (auto& [ch, freq] : frequencies) {
        minHeap.push_back(new HuffmanNode(ch, freq));
    } // @step:initialize

    // Sort ascending to simulate a min-heap
    sort(minHeap.begin(), minHeap.end(), [](HuffmanNode* nodeA, HuffmanNode* nodeB) {
        return nodeA->freq < nodeB->freq;
    }); // @step:select-min-freq

    while (minHeap.size() > 1) {
        // Extract two minimums
        HuffmanNode* leftNode = minHeap.front(); minHeap.erase(minHeap.begin()); // @step:select-min-freq
        HuffmanNode* rightNode = minHeap.front(); minHeap.erase(minHeap.begin()); // @step:select-min-freq

        // Merge into a new internal node
        HuffmanNode* merged = new HuffmanNode(leftNode->freq + rightNode->freq, leftNode, rightNode); // @step:build-node

        // Re-insert maintaining sorted order
        auto insertPos = find_if(minHeap.begin(), minHeap.end(), [&](HuffmanNode* node) {
            return node->freq > merged->freq;
        });
        minHeap.insert(insertPos, merged); // @step:build-node
    }

    HuffmanNode* huffmanRoot = minHeap.empty() ? nullptr : minHeap[0];
    map<char, string> encodings;
    generateCodes(huffmanRoot, "", encodings);
    return encodings; // @step:complete
}
